text/0000-read-all.md

```diff
@@ -0,0 +1,50 @@
+- Feature Name: read_all
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's Write trait has write_all, which attempts to write an entire
+buffer.  This proposal adds read_all, which attempts to read a fixed
+number of bytes into a given buffer.
+
+# Motivation
+
+The new read_all method will allow programs to read from disk without
+having to write their own read loops.  Most Rust programs which need
+to read from disk will prefer this to the plain read function.  Many C
+programs have the same need, and solve it the same way (e.g. git has
+read_in_full).  Here's one example of a Rust library doing this:
+https://github.com/BurntSushi/byteorder/blob/master/src/new.rs#L184
+
+# Detailed design
+
+The read_all function will take a mutable, borrowed slice of u8 to
+read into, and will attempt to fill that entire slice with data.
+
+It will loop, calling read() once per iteration and attempting to read
+the remaining amount of data.  If read returns EINTR, the loop will
+retry.  If there are no more bytes to read (as signalled by a return
+of Ok(0) from read()), a new error type, ErrorKind::ReadZero, will be
+returned.  In the event of another error, that error will be
+returned. After a read call returns having successfully read some
+bytes, the total number of bytes read will be updated.  If that
+total is equal to the size of the buffer, read will return
+successfully.
```

Comment:
> @alexcrichton 
> Technically `Ok(0)` is indeed special as cases like `Write` on `&mut [u8]` will end up generating this once you hit the end of the buffer.

Really? I would expect and strongly argue that that should generate an IO error, even for a standard `write` call, and not return `Ok(0)`. In my mind, this is analogous to a pipe or a socket with the remote end closed, where reading yields end of file, while writing results in an error. As perhaps a closer analogy to the fixed buffer case, on Linux, if I have a fixed-size block device (say a small ramdisk), and I read to the end, subsequent reads will return 0 (end of file). If, however, I reach the end of the file and try to write more, I'll receive an error (`ENOSPC`). Similarly, writing to a file that has reached the maximum allowed size results in (`EFBIG`), not a successful write of zero size.

Is there any chance this can be changed?